### PR TITLE
changed api for sorting and filtering: see PropTypes on Ardagryd. Cha…

### PR DESCRIPTION
**Pull request by fkretzer**
…nged Behaviour for sorting and filtering: Internal state is NOT OVERRIDDEN from external props, if props have not changed.

### CODE REVIEW

File: Ardagryd.js [hunk 1]
```diff
@@ -108,7 +138,16 @@ const Ardagryd = (props)=>{
         }
 
         //Sort
-        if (sortColumn){
+
+    //default order
+    var order = sortConfigFromProp(props.sort).order;
+
+    var sortColumn = sortConfigFromProp(props.sort).columnName;
```
Comment (jochenberger) [inline Ardagryd.js]: We don't want to extract the sort config twice. I'd write
```javascript
const { order, sortColumn } = sortConfigFromProp(props.sort);
```

File: Ardagryd.js [hunk 2]
```diff
@@ -288,7 +330,7 @@ class GridHeaderCell extends React.Component {
                 {label}
               </span>
               { sortable &&
-                <Button active={active} bsSize="xsmall" style={{marginLeft: '5px'}} onClick={this.sortChanged}>
+                <Button className={"sort-toggle-for-"+label} active={active} bsSize="xsmall" style={{marginLeft: '5px'}} onClick={this.sortChanged}>
```
Comment (jochenberger) [inline Ardagryd.js]: Isn't this going to cause trouble if `label` contains white-space? Why do we need a custom class here?

File: Ardagryd.js [hunk 3]
```diff
@@ -597,61 +660,60 @@ export class Grid extends React.Component {
...
     componentWillReceiveProps(nextProps){
-      this.setState({
-          config: nextProps.config ? nextProps.config : {},
-          columns: nextProps.columns ? nextProps.columns : {}
-      });
+        if (!(JSON.stringify(this.props.sort) === JSON.stringify(nextProps.sort))){
+            this.setState({sort: nextProps.sort});
+        }
+        if (!(JSON.stringify(this.props.filter) === JSON.stringify(nextProps.filter))){
+            this.setState({filter: nextProps.filter});
```
Comment (jochenberger) [inline Ardagryd.js]: It's probably more efficient to create a `newState` object with both changes combined and only call `setState` once. I'm not sure if it's advisable to only call `setState` if the object is non-empty though.

File: Ardagryd.js (hunk 2 above)
Comment (fkretzer) [inline Ardagryd.js]: good point! added it for easier testing.

File: Ardagryd.js [hunk 4]
```diff
@@ -14,6 +14,51 @@ const DESCENDING = "desc";
...
+            } else {
+                sortConfigObject = sortProp;
+            }
+            sortColumn = sortConfigObject.columnName;
+            if (sortConfigObject.order
+                && (sortConfigObject.order === ASCENDING || sortConfigObject.order === DESCENDING)){
+                order = sortConfigObject.order;
+            }
+            break;
+        default:
```
Comment (jochenberger) [inline Ardagryd.js]: Should this throw?

Comment (fkretzer) [inline Ardagryd.js]: Under which circumstances?

Comment (jochenberger) [inline Ardagryd.js]: When the `default` branch executes, i.e. if `sortProp` is neither a string nor an object/array. We could also just let is break in production. That's going to save us a few milliseconds and `propTypes` should have made sure that the user passes in valid data. Let's see if it causes trouble for anyone, we can also act on this later.

Comment (fkretzer) [inline Ardagryd.js]: ok, i then will increase the minor version because of the api change.

**Comment by jochenberger** [comment]
You could also remove the `deepmerge` dependency/import.

**Comment by jochenberger** [comment]
Alright, let's go.

**Pull request merged**